echo-notifications/src/main/groovy/com/netflix/spinnaker/echo/notification/HtmlToPlainText.java

```diff
@@ -0,0 +1,76 @@
+                                                  *
+                                                  * Licensed under the Apache License, Version 2.0 (the "License")
+                                                  * you may not use this file except in compliance with the License.
+                                                  * You may obtain a copy of the License at
+                                                  *
+                                                  *   http://www.apache.org/licenses/LICENSE-2.0
+                                                  *
+                                                  * Unless required by applicable law or agreed to in writing, software
+                                                  * distributed under the License is distributed on an "AS IS" BASIS,
+                                                  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+                                                  * See the License for the specific language governing permissions and
+                                                  * limitations under the License.
+                                                  */
+
+import org.jsoup.helper.StringUtil;
+import org.jsoup.nodes.Element;
+import org.jsoup.nodes.Node;
+import org.jsoup.nodes.TextNode;
+import org.jsoup.select.NodeTraversor;
+import org.jsoup.select.NodeVisitor;
+
+public class HtmlToPlainText {
+  public String getPlainText(Element element) {
+    FormattingVisitor formatter = new FormattingVisitor();
+    NodeTraversor traversor = new NodeTraversor(formatter);
+    traversor.traverse(element);
+    return formatter.toString();
+  }
+
+  // the formatting rules, implemented in a breadth-first DOM traverse
+  private static class FormattingVisitor implements NodeVisitor {
+    private StringBuilder accum = new StringBuilder(); // holds the accumulated text
+
+    // hit when the node is first seen
+    public void head(Node node, int depth) {
+      String name = node.nodeName();
+      if (node instanceof TextNode) {
+        append(((TextNode) node).text()); // TextNodes carry all user-readable text in the DOM.
+      } else if (name.equals("li")) {
+        append("\n * ");
+      } else if (name.equals("dt")) {
+        append("  ");
+      } else if (StringUtil.in(name, "p", "h1", "h2", "h3", "h4", "h5", "tr")) {
+        append("\n");
+      }
+    }
+
+    // hit when all of the node's children (if any) have been visited
+    public void tail(Node node, int depth) {
+      String name = node.nodeName();
+      if (StringUtil.in(name, "br", "dd", "dt", "p", "h1", "h2", "h3", "h4", "h5")) {
+        append("\n");
+      } else if (name.equals("a")) {
+        append(String.format(" <%s>", node.absUrl("href")));
+      }
+    }
+
+    // appends text to the string builder
+    private void append(String text) {
+      if (text.equals(" ")
```

Comment:
may want to clean this up so its easier to read. Can do so in two ways: 

1. Make a function for the test 
2. Add parentheses so the order of operations is easier to read without thinking too hard about :)

eg:
```
if ( text_and_accum_empty || just_newlines) {
  return;
}
```

Reply:
ah I see the left paren on the or clause, spotless messed this up!